Automattic/woocommerce-payments
# RPP: Working Data

Fixes #6688. Blocked by https://github.com/Automattic/woocommerce-payments/pull/7318

#### Changes proposed in this Pull Request

This PR makes the `PaymentContext` class functional, and adds its basic usage throughout the payment process.

- Order Service:
	- Add the `get_order_service()` method to `WC_Payments` to expose the legacy `WC_Payments_Order_Service` instance, which is also added to the `LegacyContainer`.
	- Create a new `WCPay\Internal\Service\OrderService` class. For now, we should try to keep this class from accepting `WC_Order` objects as parameters, leaning on order IDs instead. Initially, it will just have the right method signatures, but in the future we should migrate everything from `WC_Payments_Order_Service` to it. Currently, the service only contains a single method, which should be used as an example to follow throughout the payment process re-engineering.
	- Introduce a new `WCPay\Internal\DependencyManagement\ServiceProvider\GenericServiceProvider`, which will provide all generic services.
- Saved payment method: The `SavedPaymentMethod` class now stores the IDs of both the payment method, and associated token. This contrasts with the `WC_Payment_Token` instance that it used to wrap. That's done to make the object plain and serializable, forcing delegation to a service in the future, instead of using methods directly.
- Payment Request class:
	- The class is modified to provide a token ID instead of an object to the `SavedPaymentMethod` constructor (see the previous point).
	- A new `populate_context` method was introduced, which will transfer all necessary details into the `PaymentContext` class. For now, the only needed detail is the `PaymentMethodInterface` instance, and further details should be added when needed.
- `PaymentContext` class:
	- The class receives a set of private `set` and `get` methods, which will store the data in an array. This is because in the future, those methods will contain the necessary logic for logging.
	- There are a few setters and getters, showcasing how they will be used.
- The `PaymentProcessingService` is changed to accommodate all changes.
- `InitialState` is the link between all data:
	- It uses `get_order_id()` and `get_payment_method()` from the context.
	- It uses the `OrderService` to attach the payment method to the order.


#### Testing instructions

Try checking out with the new payment process enabled, and check the order meta. `_payment_method_id` should contain a payment method ID.

As always, everything is covered with tests. There's just one place where feedback is required, and I've added a comment to it.

-------------------

- [x] Covered with tests (100% coverage of `src`)
- [x] Psalm has no warnings
